### PR TITLE
[Enhancement] improve statistics strategy for insert overwrite

### PR DESCRIPTION
**Pull request by murphyatwork**
## Why I'm doing:

`INSERT OVERWRITE` would create a new partition with different id, and with version=2, which breaks the hypothesis of current load-triggered statistics collection.

## What I'm doing:
1. INSERT OVERWRITE should not re-collect the statistics but copy the existing statistics instead.
2. if changed ratio of a table/partition is less than `10%`, we don't trigger the collection

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/InMemoryStatisticStorage.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import com.starrocks.catalog.Partition;
+import com.starrocks.catalog.Table;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class InMemoryStatisticStorage implements StatisticStorage {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/InMemoryStatisticStorage.java:27]: better to move it to the test dir?

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java [hunk 2]
```diff
@@ -2063,6 +2063,10 @@ public class Config extends ConfigBase {
     @ConfField(mutable = true)
     public static long statistic_sample_collect_rows = 200000;
 
+    @ConfField(mutable = true, comment = "If changed ratio of a table/partition is larger than this threshold, " +
+            "we would use sample statistics instead of full statistics")
+    public static double statistic_sample_collect_ratio_threshold = 0.1;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/common/Config.java:2068]:
```suggestion
    public static double statistic_sample_collect_ratio_threshold_on_first_load = 0.1;
```

File: fe/fe-core/src/main/java/com/starrocks/statistic/StatisticsCollectionTrigger.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        analyzeType = decideAnalyzeTypeForOverwrite();
+    }
+
+    private StatsConstants.AnalyzeType decideAnalyzeTypeForOverwrite() {
+        Preconditions.checkNotNull(overwriteJobStats, "must have overwrite stats");
+        long sourceRows = overwriteJobStats.getSourceRows();
+        long targetRows = overwriteJobStats.getTargetRows();
+        double deltaRatio = 1.0 * (targetRows - sourceRows) / (sourceRows + 1);
+        if (deltaRatio < Config.statistic_sample_collect_ratio_threshold) {
+            return null;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/statistic/StatisticsCollectionTrigger.java:267]: shuold be `FULL` ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/InMemoryStatisticStorage.java (hunk 1 above)
Comment (murphyatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/InMemoryStatisticStorage.java]: fixed

File: fe/fe-core/src/main/java/com/starrocks/common/Config.java (hunk 2 above)
Comment (murphyatwork) [reply fe/fe-core/src/main/java/com/starrocks/common/Config.java]: fixed

**Review by packy92 — DISMISSED** [review]

**Review by Seaven — DISMISSED** [review]

**Comment by murphyatwork** [comment]
@Mergifyio rebase

**Review by Seaven — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Comment by wangsimo0** [comment]
ignore backport check: 3.3.5

**Pull request merged**